### ISSUE TITLE
RP-1 Roadmap - Updated 5-14-2019

### ISSUE THREAD

**Issue by pap1723** [KSP-RO/RP-0#1087]
Obviously we did a terrible time of sticking to the previous roadmap, but that was due to a number of factors. We are currently at a crossroads here. We have a version for 1.3.1 that is definitely good enough to release right now. BUT, we run into the problem of requiring so many backports, patches and copies of mods that it becomes essentially impossible to release through CKAN. I am up for ideas on how to proceed, here is what I see are our options:

1. Release RP-1 for 1.3.1 as is on CKAN. Have detailed install instructions on how to patch the necessary mods to get it working properly.
2. Leave RP-1 for 1.3.1 as is. Create a branch and release a Github version of what it currently is and provide install instructions.
3. Try to figure out a way to combine the needed patches, mods and backports into a "mod pack" and release RP 1.3.1.

## SHORT-TERM GOALS

- [ ] Change the KCT BP Formula to Root Based One
- [ ] Add Contracts for After First Science Satellite Before Solar Satellite #990 
- [ ] Balance Payouts on Satellite Contracts #990 
- [ ] Procedural Avionics Probe Core Size Issues vs Prefab Ones #1081 #968 
- [ ] 1.6+ Release

## INTERMEDIATE GOALS

- [ ] Integrating all Tank Types into One Tank
- [ ] MLI Balance Changes (limiting MLI by tank type)
- [ ] Add Turbopump ModuleTag to Nuclear Engines for cost balance #1026 
- [ ] DOCUMENTATION!

## LONG-TERM GOALS

- [ ] Prefab Tanks Tooling #962 
- [ ] KCT Launch Costs Fixes #965 
- [ ] RealBattery Design

**Comment by pap1723** [comment]
Input please: @kurgut @siimav @leudaimon @theysen @MikeOnTea @winged7 @Bornholio

**Comment by leudaimon** [comment]
Given how far you are in the 1.6+ side, I would say going with option 2) is totally fine, it will be too much work to try to get the other alternatives going.

About the roadmap itself, I think it's very good. One thing I would add in the MLI changes, would be to think about some way of having something similar to the Centaur and/or Shuttle ET, which gives insulation and is also resistant to the atmospheric heat during launch. Explodey tanks with 2-3 MLI is not fun. I also don't understand the item about contracts after FSO and before First Solar. First solar is usually completed before FSO, as you need the electronics node with the solar panels to unlock the experiment for FSO.

**Comment by pap1723** [comment]
@leudaimon thanks, added the MLI note as I agree.

Did we change that at some point? It didn't used to require it, so that is probably a moot point now.

**Comment by leudaimon** [comment]
yeah, we started requiring experiments in the satellite era science node instead of the first science node, exactly to make FSO more difficult. I still think it could be interesting to include a couple more contracts besides polar that you can complete after FO and before Solar/FSO. One of them could be a Sputnik 2, requiring a bio sample in orbit (poor Laika), giving some science in addition to money. The science part wouldn't be necessary with kerbalism if the duration of the bio sample experiment is longer than a SR can achieve.

**Comment by PhineasFreak** [comment]
We really need to separate the RF insulation into two categories, SOFI (light, can be used externally but can only keep the cryogenic contents for a short period of time) and MLI (heavy, can only be used in space but can keep the cryogenic contents indefinitely with active cooling).

That of course is a job of RF (make these internal parameters accessible via config files).

**Comment by glilienthal** [comment]
coming back to the original question: I'd really like to encourage you all to go for option 1:  **Release RP-1 for 1.3.1 as is on CKAN. Have detailed install instructions on how to patch the necessary mods to get it working properly.**

This would give a lot more people to confidence to try it, even if you have to do some manual installs. 

I think as it currently stands, RP-1 is great, and adds so much to the game. It's totally worth a lot of people playing. (At least I enjoy it immensely.) and giving it a "release" tag would inspire so much more well-earned confidence.

**Comment by norcalplanner** [comment]
And the new release version needs a suitable version number, 1.1 or higher.

**Comment by 7ranceaddic7** [comment]
Hi guys.  I'm not someone who currently uses RP-1, but I pay attention to the development.  As an "outsider",  I'd have to agree with @glilienthal regarding an "As Is" release.

You've all been doing a great deal of development toward 1.6.x.  The user-base for RP-1 is most likely sufficiently advanced that they can deal with a detailed instruction list.  Any attempt to backport a release for 1.3.1 is most probably a large a waste of time and effort that could be used to bring RP-1 advance a 1.6.x release.

I hate seeing good projects get bogged-down.  So, as that "outsider", I'd be willing to be the uninitiated slob who will test those install instructions.  If I've never used RP-1 and can install according to instructions then play a couple of hours ... I think you've got your solution.

**Comment by Bornholio** [comment]
I'd vote for straight to 1.6.x and love the list. maybe i'll get to play in the next year. .sigh

**Comment by robinxb** [comment]
As a player, watching you guys working for more than two years, I would love to say, keep going and leave whatever 1.3.1 is now. 
I'm willing to upgrade my ksp version whenever you release the first version of 1.6.x

One last and important thing is,  thank you all for making these mods.

**Comment by sneakertack** [comment]
As a player, having come across https://github.com/KSP-RO/RP-0/releases/tag/v1.1 and https://github.com/KSP-RO/RP-0/wiki/RP-1-v1.1-Installation-Instructions, I was under the impression that the team must have had gone ahead with Option 1. I then realized that:

1. On CKAN, the mod version of RP-1 is still v1.00
2. RP-1-v1.1.zip seems to be a corrupted(?) zip file (fails to unzip on Ubuntu 18.04 (Archive Manager or `unzip` on CLI))

Is it correct to assume that Option 1 is in the midst of happening?

**Comment by glilienthal** [comment]
There is no https://github.com/KSP-RO/RP-0/releases/tag/v1.1.1 , so I think option 1 is done! 

Thanks to all!

**Comment by Bornholio** [comment]
Added an install 1.6.1 version wiki, needs work as per usual https://github.com/KSP-RO/RP-0/wiki/Installation-for-1.6.1-(Very-WIP)

**Comment by zajc3w** [comment]
Just an input from massive fan.
Thank You for greatest set of mods ever!!
Could you move to 1.7.1? Just for sake of having working robotics please.
I folowed 1.6.1 installation instructions and can't see any problems in 1.7.1 compared to 1.6.1.
Recompiled RSS, Kopernicus, real heat and deadly reentry. Flew couple missions each, and apart from 1.3.1 saves loosing half of my ships due to missing probe core(RO_early_probe_core), science parts not having experiments assigned and first RTG not generating electricity, everything looks good and the same in both installs.
Kopernicus is updated since my tests, so move to 1.7.1 seems to me(no idea about making mods really) to be relatively easy.
I'm just sick of Infernal Robotics breaking ships.
If not I'll just bodge my install on 1.71 ;)

**Comment by siimav** [comment]
Both ` Integrating all Tank Types into One Tank` and ` MLI Balance Changes (limiting MLI by tank type)` have been implemented but will probably need the modified RF dll from ROTanks. There's a PR to the RF repo that will need to be merged before those 2 can be released for public use.
Should we still do some additional cost or mass balancing for MLI? Currently 30 layers of MLI will add rougly 10% cost extra for the tank and the mass difference is almost nothing. According to Starwaster the current MLI layer cost contains both the cost of materials and also a guesstimate on installation. Maybe those tradeoffs are still too small for almost completely eliminating boiloff?

**Comment by MikeOnTea** [comment]
Currently adding too much MLI risks overheating/explosion of atmospheric stages, not sure if that's a bug or a feature. ;) Also not sure if we might want to have 2 insulation options, foam for atmospheric stages and MLI for space stages only?

**Comment by Gordon-Dry** [comment]
Choosing the "foam" mesh at ROTanks could also be atmospheric insulation. Why not?

**Comment by siimav** [comment]
I think the fact that MLI causes overheating on tanks is half bug, half feature. It's somewhat realistic that the skin of the tank starts overheating if the insulation prevents transferring the heat into the interior of the tank. The strange part is that this issue only happens at some specific speed and altitude combinations. 
Having 2 different insulation options sounds good to me but I guess that would need quite extensive changes to RF code. I think internally all the tank insulation is handled as various amounts of MLI layers. Tagging @Starwaster in case he wants to chime in here.

**Comment by ec429** [comment]
I vote for special-casing mli=1 to mean SOFI.  Of course we'd want to make the GUI indicate this, rather than just requiring people to know.

**Comment by Bornholio** [comment]
From my understanding of MLI, using 3-10 layers is likely, less is not MLI and more is diminishing returns.  As it is currently it uses the empirical formula. My opinion is that tanks that can use it be limited to Tank III and Tank IV.  Insulated (foam or otherwise) tank could be represented by 1-2 layers but it won't be heavy enough https://dura-foam.com/resources/foam-roofing/nasa-shuttle-fuel-tank/ on the SS-ET about 2 of the ~30-35 tons are insulation, no idea what to do about overheating other than sanely using the stuff. I count it as part of the learning curve. Perhaps we need a link to a page in RF Wiki that explains how to safely use it and why and when it is reasonable to use.  I think it is properly realistic assuming you use it sensibly.

**Comment by Starwaster** [comment]
Re: MLI. Where to begin. First MLI vs SOFI: SOFI is handled in TANK_DEFINITION in the invidivual TANK nodes. Nearly all cryogenic resource TANK nodes are configured for SOFI except in the Default tank which was _originally_ (pre-RO) configured according what was historically present on tanks of the Saturn V era. Not what was technically possible during that era but what was actually done historically. SOFI weight is done by setting weight on the TANK appropriately. The hydrogen tank had the correct values for foam. 

MLI is lighter than SOFI not heavier. Of course that's going by realism not 'game balance'. I will always fall on the side of realism vs game balance which is by comparison arbitrary.

Overheating MLI: I assume you're talking on ascent? The solution is no solution. Do nothing. It's not your job as RO developers. It's the job of players to construct sensibly. Either to not put MLI where it can be lost or to put it in fairings. If you want an RF solution then it can be made destructible which would be realistic. MLI shouldn't be exposed to overheating conditions and should melt and burn off. I.e. it would lose layers. As layers are lost the skin-internal conductivity would be increased transfering heat into the interior and overheating would stop but insulation loss would be permanent. This is something I would be in favor of but it involves adding a new feature and currently I am focused on fixing bugs rather than adding features. Unnecessary features even less so.

And again let me stressthat there is no need to 'simulate' SOFI through MLI value munging. It is already present and has been ever since RF's boiloff was rewritten for KSP 1.0

**Comment by Starwaster** [comment]
I also have to add (and I apologize if I come off cranky in this) that the addition to per-tank limits on MLI are unnecessary and unrealistic. It's resulted in an what I consider unnecessary feature requests to RF and unnecessary pull requests; especially when bug fixing should be the priority.

MLI limitation should (and currently CAN be without any unnecessary coding) be global. It's already possible to use the tech tree system to increase the limit on MLI via ModuleFuelTanks. If you can add 60 layers to a tank you can add 60 layers to any tank. Let the player decide based on the global limit.

**Comment by siimav** [comment]
While you might be right that MLI limitations should be implemented directly through the tech tree, I disagree with you saying that the PR adds *an unnecessary feature*. Besides MLI, it also added fields for minUtilization and maxUtilization which are absolutely crucial for being able to unify the current Tank-I through Tank-IV into a single tank part. It's not only my opinion that this change would be a huge help for any RP-1 player.

**Comment by MikeOnTea** [comment]
Making MLI disintegrate on ascent would probably be nice to have in the long run, but we'd need to be able to use SOFI on tanks then - either through Cryo-Tank-1/2/3/4 types or by having a slider like MLI which would probably be more user-friendly.

**Comment by Starwaster** [comment]
@siimav hard limits aren't really helping the player. minUtilization doesn't even make sense. In fact it's antithetical to what MFS is all about.

@MikeOnTea Making the other insulation system player configurable would definitely have to be a long term thing then since it would require moving the configuration data out of TANK_DEFINITION and onto ModuleFuelTanks

**Comment by ec429** [comment]
> And again let me stress that there is no need to 'simulate' SOFI through MLI value munging. It is already present and has been ever since RF's boiloff was rewritten for KSP 1.0.

So just to be clear, if you take an LH₂ tank and add MLI, does it have (and pay the weight penalty for) **both** MLI and the built-in SOFI?  And is that what we want?

**Comment by Starwaster** [comment]
@ec429 MLI mass and cost are calculated separately based on tank area and number of layers and are in addition to all other tank cost and and mass as determined by the tank definition (TANK_DEFINITION)
